src/RadixDlt.NetworkGateway.PostgresIntegration/Models/EntityResourceVaultAggregateHistory.cs

```diff
@@ -84,6 +89,107 @@ internal class EntityResourceVaultAggregateHistory
+    }
+
+    /// <summary>
+    /// Attempts to add new or update existing vault unless it already is the most recently modified one.
+    /// </summary>
+    /// <returns>true if added or modified, otherwise false.</returns>
+    public bool TryUpsert(long vaultId, long stateVersion)
+    {
+        EnsureVaultSignificantUpdateStateVersionsExists();
+
+        var currentIndex = VaultEntityIds.IndexOf(vaultId);
+
+        // we're already the most recent one and there's no point potential second resource with the same state version
+        if (currentIndex == 0 && VaultEntityIds.Count == 1)
+        {
+            return false;
+        }
+
+        // we're already the most recent one but there might be second resource with matching state_version
+        // so in order to guarantee that we'll be unambiguously earlier in the overall sequence we might potentially
+        // want to update anyways
+        if (currentIndex == 0 && VaultEntityIds.Count > 1 && VaultSignificantUpdateStateVersions[1] < VaultSignificantUpdateStateVersions[0])
+        {
+            return false;
+        }
+
+        if (currentIndex != -1)
+        {
+            VaultEntityIds.RemoveAt(currentIndex);
+            VaultSignificantUpdateStateVersions.RemoveAt(currentIndex);
+        }
+
+        VaultEntityIds.Insert(0, vaultId);
+        VaultSignificantUpdateStateVersions.Insert(0, stateVersion);
+
+        return true;
+    }
+
+    public bool ShouldBePersisted()
+    {
+        if (_originalVaultEntityIds == null)
+        {
+            return true;
+        }
+
+        if (!_originalVaultEntityIds.SequenceEqual(VaultEntityIds.ToArray()))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    private static EntityResourceVaultAggregateHistory CopyOrCreate(long databaseId, EntityResourceVaultAggregateHistory? other, long entityId, long resourceEntityId, long stateVersion)
+    {
+        var ret = new EntityResourceVaultAggregateHistory
+        {
+            Id = databaseId,
+            FromStateVersion = stateVersion,
+            EntityId = entityId,
+            ResourceEntityId = resourceEntityId,
+            VaultEntityIds = new List<long>(other?.VaultEntityIds.ToArray() ?? Array.Empty<long>()),
+            VaultSignificantUpdateStateVersions = new List<long>(other?.VaultSignificantUpdateStateVersions?.ToArray() ?? Array.Empty<long>()),
+        };
+
+        if (other != null)
+        {
+            ret._originalVaultEntityIds = ret.VaultEntityIds.ToArray();
+
+            ret.EnsureVaultSignificantUpdateStateVersionsExists();
+        }
+
+        return ret;
+    }
+
+    [MemberNotNull(nameof(VaultSignificantUpdateStateVersions))]
+    private void EnsureVaultSignificantUpdateStateVersionsExists()
+    {
+        VaultSignificantUpdateStateVersions ??= new List<long>(Enumerable.Repeat(1L, VaultEntityIds.Count));
```

Comment:
I wonder if we should eg use -1L or something here, to denote placeholder, rather than be confused for the actual state version 1?